src/trainer.py

```diff
@@ -0,0 +1,222 @@
+        self.lr_scheduler_disc_d = torch.optim.lr_scheduler.LambdaLR(
+            self.optimizer_disc_deshadower,
+            lr_lambda=LR_lambda(opt.n_epochs, opt.epoch, opt.decay_epoch).step,
+        )
+
+    def run_one_batch_for_generator(
+        self, real_shadow, real_mask, mask_non_shadow, mask_queue, target_real
+    ):
+        pass
+        # TODO finish
+        # zero_grad()
+        self.optimizer_gen.zero_grad()
+
+        same_mask = self.generator_shadow_to_free(real_shadow)
+        same_shadow = self.generator_free_to_shadow(real_mask, mask_non_shadow)
+
+        # Identity loss
+        identity_loss_mask = self.identity_loss_criterion(same_mask, real_mask) * 5.0
+        identity_loss_shadow = (
+            self.identity_loss_criterion(same_shadow, real_shadow) * 5.0
+        )
+
+        # GAN loss
+        fake_mask = self.generator_shadow_to_free(real_shadow)
+        pred_fake = self.discriminator_free_to_shadow(fake_mask)
+
+        loss_gen_shadow_to_free = self.gan_loss_criterion(pred_fake, target_real)
+        mask_queue.insert(mask_generator(real_shadow, fake_mask))
+
+        fake_shadow = self.generator_free_to_shadow(real_mask, mask_queue.rand_item())
+        pred_fake = self.discriminator_shadow_to_free(fake_shadow)
+        loss_gen_free_to_shadow = self.gan_loss_criterion(pred_fake, target_real)
+
+        # Cycle loss
+        recovered_shadow = self.generator_free_to_shadow(
+            fake_mask, mask_queue.last_item()
+        )
+        loss_cycle_shadow = (
+            self.cycle_loss_criterion(recovered_shadow, real_shadow) * 10.0
+        )
+
+        recovered_mask = self.generator_shadow_to_free(fake_shadow)
+        loss_cycle_mask = self.cycle_loss_criterion(recovered_mask, real_mask) * 10.0
+
+        # Total loss
+        gen_loss = (
+            identity_loss_shadow
+            + identity_loss_mask
+            + loss_gen_shadow_to_free
+            + loss_gen_free_to_shadow
+            + loss_cycle_shadow
+            + loss_cycle_mask
+        )
+        gen_loss.backward()
+
+        self.optimizer_gen.step()
+
+    def run_one_batch_for_discriminator_s2f(
+        self,
+        real_shadow,
+        real_mask,
+        target_real,
+        target_fake,
+        fake_shadow_buff,
+        mask_queue,
+    ):
+        # zero_grad()
+        self.optimizer_disc_deshadower.zero_grad()
+
+        # Real loss
+        prediction_real = self.discriminator_shadow_to_free(real_shadow)
+        loss_disc_real = self.gan_loss_criterion(prediction_real, target_real)
+
+        # Fake loss
+        fake_shadow = self.generator_free_to_shadow(real_mask, mask_queue.rand_item())
+        fake_shadow = fake_shadow_buff.push_and_pop(fake_shadow)
+        prediction_fake = self.discriminator_shadow_to_free(fake_shadow.detach())
+        loss_disc_fake = self.gan_loss_criterion(prediction_fake, target_fake)
+
+        # Total loss
+        loss_disc = (loss_disc_real + loss_disc_fake) / 2.0
```

Comment:
why float instead of int

Reply:
We don't want to loose informations about losses. Especially when they update in next iterations.